lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp

```diff
@@ -119,21 +119,85 @@ class TensorEmptyConversionPattern
 
 // TTIR::ToLayoutOp is a rather generic op that dictates how all the layout
 // properties of a tensor should be set. However, in TTNN world, multiple APIs
-// are required to achieve an arbitrary layout. There are two main distinct
-// paths in this conversion pattern:
-//
-// 1. If the layout calls for device memory, we will call TTNN::ToLayoutOp and
-//    TTNN::ToDeviceOp to achieve the desired layout.
-//
-// 2. If the layout calls for system memory, we will call TTNN::ToLayoutOp to
-//    change the tensor to RowMajor layout, and then the TTNN::FromDeviceOp to
-//    move to host memory
-//
+// are required to achieve an arbitrary layout, namely:
+// - ToLayoutOp: to set the layout (ROW_MAJOR, TILE) of the tensor
+// - TypecastOp: to change the data type of the tensor
+// - ToDeviceOp: to move the tensor to a specific device
+// - FromDeviceOp: to move the tensor from a specific device to host
+// - ToMemoryConfigOp: to set the memory configuration (dram, l1, interleaved,
+// sharded) of the tensor
 class ToLayoutOpConversionPattern
     : public OpConversionPattern<ttir::ToLayoutOp> {
 public:
   using OpConversionPattern<ttir::ToLayoutOp>::OpConversionPattern;
 
+  LogicalResult
+  matchAndRewrite(ttir::ToLayoutOp op, OpAdaptor adaptor,
```

Comment:
My ask here would be to try and migrate this decomposition of ToLayoutOp to a separate TTNN pass which occurs after lowering and thus keep lowering simple and clean. This would also allow optimizer to work with single memory layout manipulation op(for example ttnn:ToLayoutOp) while a separate pass would properly decompose it.

Reply:
Added a separate pass after the optimizer that decomposes the composite toLayout op instead of doing the decomposing during the lowering stage. Updated optimizer accordingly

Reply:
@jnie-TT Looks great on a first pass, will take a more detailed look on Monday morning.

Reply:
Looks good, please proceed.